File: src/main/java/org/tdf/lotusvm/types/FunctionSection.java
package org.tdf.lotusvm.types;

import lombok.Getter;
import org.tdf.lotusvm.common.BytesReader;

/**
 * The function section has the id 3.
 * It decodes into a vector of type indices that represent the type ﬁelds of the functions in the funcs component of a module.
 * The locals and body ﬁelds of the respective functions are encoded separately in the code section.
 */
public class FunctionSection extends AbstractSection {
    @Getter
    private int[] typeIndices;

    public FunctionSection(SectionID id, long size, BytesReader payload, int offset, int limit) {
        super(id, size, payload, offset, limit);
    }

    @Override
    void readPayload() {
        typeIndices = getReader().readUint32Vec();
    }
}

File: src/main/java/org/tdf/lotusvm/types/FunctionSection.kt
package org.tdf.lotusvm.types

import org.tdf.lotusvm.common.BytesReader

/**
 * The function section has the id 3.
 * It decodes into a vector of type indices that represent the type ﬁelds of the functions in the funcs component of a module.
 * The locals and body ﬁelds of the respective functions are encoded separately in the code section.
 */
class FunctionSection(id: SectionID, size: Long, payload: BytesReader, offset: Int, limit: Int) :
    AbstractSection(id, size, payload, offset, limit) {
    var typeIndices: IntArray = IntArray(0)
        private set

    public override fun readPayload() {
        typeIndices = reader.readUint32Vec()
    }
}